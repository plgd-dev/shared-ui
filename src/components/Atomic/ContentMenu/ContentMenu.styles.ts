import { css } from '@emotion/react'
import { ThemeType, get } from '../_theme'

export const contentMenu = (theme: ThemeType) => css`
    padding: 16px 12px 12px 12px;
    border-radius: 16px;
    border: 1px solid ${get(theme, `ContentMenu.borderColor`)};
    background: ${get(theme, `ContentMenu.background`)};
    box-shadow: 0 1px 2px 0 rgba(228, 229, 231, 0.24);
`

export const title = (theme: ThemeType) => css`
    color: ${get(theme, `ContentMenu.title.color`)};
    margin-bottom: 12px;
    padding: 0 12px;
`

export const searchWrapper = css`
    position: relative;
`

export const searchIcon = css`
    position: absolute;
    top: 15px;
    left: 12px;
`

export const search = (theme: ThemeType) => css`
    background: ${get(theme, `ContentMenu.search.background`)};
    border: none;
    padding: 12px 12px 12px 40px;
    font-size: 14px;
    font-style: normal;
    font-weight: 400;
    line-height: 22px;
    color: ${get(theme, `ContentMenu.search.color`)};
    width: 100%;
    margin-bottom: 12px;
    border-radius: 8px;

    &:focus {
        outline: none;
        border: none;
    }
`

export const menuList = css`
    list-style: none;
    padding: 0;
    margin: 0;

    .item-enter {
    }

    .item-enter-done {
        max-height: 200px;
    }

    .item-enter-active {
        max-height: 200px;
    }

    .item-exit {
        max-height: 0;
    }

    .item-exit-active {
        max-height: 0;
    }
`

export const itemTitleIcon = css`
    flex: 0 0 16px;
`

export const itemTitleText = css`
    display: flex;
    align-items: center;
    margin-left: 12px;
    font-weight: bold;
`

export const titleNoMargin = css`
    margin-left: 0;
`

export const item = (theme: ThemeType) => css`
    display: flex;
    align-items: center;
    padding: 10px 12px;
    text-decoration: none;
    min-height: 40px;
    box-sizing: border-box;
    border-radius: 8px;
    color: ${get(theme, `ContentMenu.item.color`)};
    transition: all 0.25s;
    position: relative;
    overflow: hidden;

    &:hover {
        text-decoration: none !important;
        color: ${get(theme, `ContentMenu.item.hover.color`)};

        .icon {
            color: ${get(theme, `ContentMenu.item.hover.icon.color`)};
        }
    }
`

export const activeItem = (theme: ThemeType) => css`
    color: ${get(theme, `ContentMenu.item.active.color`)};
    background: ${get(theme, `ContentMenu.item.active.background`)};

    .icon {
        color: ${get(theme, `ContentMenu.item.active.icon.color`)}!important;
    }

    &:hover {
        text-decoration: none !important;
        color: ${get(theme, `ContentMenu.item.active.hover.color`)};
    }
`

export const arrow = css`
    position: absolute;
    right: 8px;
    top: 50%;
    transform: translateY(-50%);
    width: 10px;
    height: 10px;
    display: flex;
    align-items: center;
    justify-content: center;
    transition: all 0.25s;
`

export const activeArrow = css`
    transform: translateY(-50%) rotate(180deg);
`

export const subItems = css`
    // max-height: 0;
    overflow: hidden;
    //transition: all 0.45s;
`

export const subItemsList = css`
    margin: 0;
    padding: 10px 0 0 20px;
    list-style: none;
`

export const subItemLink = (theme: ThemeType) => css`
    display: block;
    position: relative;
    padding: 8px 8px 8px 20px;
    white-space: nowrap;
    color: ${get(theme, `ContentMenu.subItem.color`)};

    &:hover {
        color: ${get(theme, `ContentMenu.subItem.hover.color`)};
    }
`

export const subItemLinkActive = (theme: ThemeType) => css`
    color: ${get(theme, `ContentMenu.subItem.active.color`)};
`

export const line = css`
    width: 14px;
    height: 55px;
    position: absolute;
    left: 0;
    bottom: 50%;
`

export const icon = css`
    margin-left: 8px;
`

export const iconSuccess = (theme: ThemeType) => css`
    color: ${get(theme, `ContentMenu.item.icon.success.color`)};
`

export const iconError = (theme: ThemeType) => css`
    color: ${get(theme, `ContentMenu.item.icon.error.color`)};
`
