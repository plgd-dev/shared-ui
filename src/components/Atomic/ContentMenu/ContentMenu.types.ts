import { ReactNode, SyntheticEvent } from 'react'

export type ItemType = {
    children?: SubItemItem[]
    icon: ReactNode
    id: string
    link: string
    title: string
}

export type SubItemItem = Omit<ItemType, 'icon' | 'link'> & {
    contentRef?: any
}

export type Props = {
    activeItem: string
    className?: string
    handleItemClick: (item: ItemType, e: SyntheticEvent) => void
    handleSubItemClick?: (subItem: SubItemItem, item: ItemType, e: SyntheticEvent) => void
    id?: string
    menu: ItemType[]
    title: string
}
