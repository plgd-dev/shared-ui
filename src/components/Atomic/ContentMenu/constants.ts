export const statuses = {
    SUCCESS: 'success',
    ERROR: 'error',
    WARNING: 'warning',
} as const
