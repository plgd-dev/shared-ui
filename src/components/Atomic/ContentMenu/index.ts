export { default } from './ContentMenu'
export * from './ContentMenu'
