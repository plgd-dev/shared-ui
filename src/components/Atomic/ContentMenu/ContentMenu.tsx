import React, { cloneElement, FC, ReactElement, useEffect, useState } from 'react'
import { AnimatePresence, motion } from 'framer-motion'
import isFunction from 'lodash/isFunction'

import * as styles from './ContentMenu.styles'
import { ItemType, Props } from './ContentMenu.types'
import { Icon } from '../Icon'
import { Arrow } from '../../Layout/LeftPanel/components'
import img from '../../Layout/LeftPanel/assets/line.png'
import IconSearch from '../Icon/components/IconSearch'
import { statuses } from '../Tabs/constants'
import IconCheck from '../Icon/components/IconCheck'
import IconWarningCircle from '../Icon/components/IconWarningCircle'

const ContentMenu: FC<Props> = (props) => {
    const { activeItem, className, handleItemClick, handleSubItemClick, id, menu: menuProp, menuSearch, title } = props

    const [search, setSearch] = useState('')
    const [menu, setMenu] = useState(menuProp)

    const getIcon = (item: ItemType) => {
        if (!item.icon) {
            return null
        }

        if (typeof item.icon === 'string') {
            return <Icon icon={item.icon} />
        } else {
            return cloneElement(item.icon as ReactElement, {
                css: [styles.itemTitleIcon],
                className: 'icon',
            })
        }
    }

    useEffect(() => {
        setMenu(menuProp.filter((item) => item.title.toLocaleLowerCase().includes(search.toLocaleLowerCase())))
    }, [search, menuProp])

    return (
        <div className={className} css={styles.contentMenu} id={id}>
            {title && <div css={styles.title}>{title}</div>}
            {menuSearch && (
                <div css={styles.searchWrapper}>
                    <span css={styles.searchIcon}>
                        <IconSearch />
                    </span>
                    <input css={styles.search} onChange={(e) => setSearch(e.target.value)} role='search' type='search' value={search} />
                </div>
            )}
            <ul css={styles.menuList}>
                {menu.map((item, key) => {
                    const isItemActive = item.id === activeItem || item?.children?.some((subItem) => subItem.id === activeItem)
                    return (
                        <li key={key}>
                            <a
                                css={[styles.item, isItemActive && styles.activeItem]}
                                href='#'
                                onClick={(e) => {
                                    e.preventDefault()
                                    handleItemClick(item, e)
                                }}
                            >
                                {getIcon(item)}
                                <span aria-label={item.title} css={[styles.itemTitleText, !item.icon && styles.titleNoMargin]}>
                                    {item.title}
                                    {item.status === statuses.SUCCESS && <IconCheck css={[styles.icon, styles.iconSuccess]} />}
                                    {item.status === statuses.ERROR && <IconWarningCircle css={[styles.icon, styles.iconError]} />}
                                </span>
                                {item.children && (
                                    <span css={[styles.arrow, isItemActive && styles.activeArrow]}>
                                        <Arrow height={6} width={10} />
                                    </span>
                                )}
                            </a>
                            <AnimatePresence initial={false}>
                                {item.children && isItemActive && (
                                    <motion.section
                                        animate='open'
                                        css={styles.subItems}
                                        exit='collapsed'
                                        initial='collapsed'
                                        key='content'
                                        transition={{ duration: 0.8, ease: [0.04, 0.62, 0.23, 0.98] }}
                                        variants={{
                                            open: { opacity: 1, height: 'auto' },
                                            collapsed: { opacity: 0, height: 0 },
                                        }}
                                    >
                                        <ul css={styles.subItemsList}>
                                            {item.children?.map((subItem, key) => (
                                                <li key={key}>
                                                    <a
                                                        css={[styles.subItemLink, subItem.id === activeItem && styles.subItemLinkActive]}
                                                        href='#'
                                                        onClick={(e) => {
                                                            e.preventDefault()
                                                            handleSubItemClick &&
                                                                isFunction(handleSubItemClick(subItem, item, e)) &&
                                                                handleSubItemClick(subItem, item, e)
                                                        }}
                                                    >
                                                        <img alt='line' css={styles.line} src={img} />
                                                        {subItem.title}
                                                    </a>
                                                </li>
                                            ))}
                                        </ul>
                                    </motion.section>
                                )}
                            </AnimatePresence>
                        </li>
                    )
                })}
            </ul>
        </div>
    )
}

ContentMenu.displayName = 'ContentMenu'

export default ContentMenu
