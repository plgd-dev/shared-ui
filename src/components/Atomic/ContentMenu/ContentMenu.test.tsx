import { render, fireEvent } from '@testing-library/react'
import ContentMenu from './ContentMenu'

describe('<ContentMenu>', () => {
    const mockHandleItemClick = jest.fn()
    const mockHandleSubItemClick = jest.fn()

    const defaultProps = {
        activeItem: 'item1',
        handleItemClick: mockHandleItemClick,
        handleSubItemClick: mockHandleSubItemClick,
        id: 'test-menu',
        menu: [
            {
                id: 'item1',
                link: 'item1',
                title: 'Item 1',
                children: [
                    {
                        id: 'subitem1',
                        title: 'Subitem 1',
                    },
                ],
            },
            {
                id: 'item2',
                link: 'item2',
                title: 'Item 2',
            },
        ],
    }

    it('renders without crashing', () => {
        const { asFragment } = render(<ContentMenu {...defaultProps} />)
        expect(asFragment()).toMatchSnapshot()
    })

    it('calls handleItemClick when an item is clicked', () => {
        const { getByText } = render(<ContentMenu {...defaultProps} />)
        fireEvent.click(getByText('Item 1'))
        expect(mockHandleItemClick).toHaveBeenCalled()
    })

    it('calls handleSubItemClick when a subitem is clicked', () => {
        const { getByText } = render(<ContentMenu {...defaultProps} />)
        fireEvent.click(getByText('Subitem 1'))
        expect(mockHandleSubItemClick).toHaveBeenCalled()
    })

    it('does not call handleSubItemClick when a subitem is clicked but handleSubItemClick is not provided', () => {
        const { getByText } = render(<ContentMenu {...defaultProps} handleSubItemClick={undefined} />)
        fireEvent.click(getByText('Subitem 1'))
        expect(mockHandleSubItemClick).not.toHaveBeenCalled()
    })

    it('filters menu items based on search input', () => {
        const { getByText, getByRole } = render(<ContentMenu {...defaultProps} menuSearch />)
        fireEvent.change(getByRole('search'), { target: { value: 'Item 2' } })
        expect(getByText('Item 2')).toBeInTheDocument()
        expect(() => getByText('Item 1')).toThrow()
    })
})
