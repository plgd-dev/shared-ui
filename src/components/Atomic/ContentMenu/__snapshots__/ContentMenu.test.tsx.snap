// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentMenu> renders without crashing 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding: 16px 12px 12px 12px;
  border-radius: 16px;
  border: 1px solid;
  box-shadow: 0 1px 2px 0 rgba(228, 229, 231, 0.24);
}

.emotion-1 {
  list-style: none;
  padding: 0;
  margin: 0;
}

.emotion-1 .item-enter-done {
  max-height: 200px;
}

.emotion-1 .item-enter-active {
  max-height: 200px;
}

.emotion-1 .item-exit {
  max-height: 0;
}

.emotion-1 .item-exit-active {
  max-height: 0;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 12px;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 40px;
  box-sizing: border-box;
  border-radius: 8px;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  position: relative;
  overflow: hidden;
}

.emotion-2:hover {
  -webkit-text-decoration: none!important;
  text-decoration: none!important;
}

.emotion-2 .icon {
  color: !important;
}

.emotion-2:hover {
  -webkit-text-decoration: none!important;
  text-decoration: none!important;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 12px;
  font-weight: bold;
  margin-left: 0;
}

.emotion-4 {
  position: absolute;
  right: 8px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: 10px;
  height: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  -webkit-transform: translateY(-50%) rotate(180deg);
  -moz-transform: translateY(-50%) rotate(180deg);
  -ms-transform: translateY(-50%) rotate(180deg);
  transform: translateY(-50%) rotate(180deg);
}

.emotion-5 {
  overflow: hidden;
}

.emotion-6 {
  margin: 0;
  padding: 10px 0 0 20px;
  list-style: none;
}

.emotion-7 {
  display: block;
  position: relative;
  padding: 8px 8px 8px 20px;
  white-space: nowrap;
}

.emotion-8 {
  width: 14px;
  height: 55px;
  position: absolute;
  left: 0;
  bottom: 50%;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 12px;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 40px;
  box-sizing: border-box;
  border-radius: 8px;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  position: relative;
  overflow: hidden;
}

.emotion-9:hover {
  -webkit-text-decoration: none!important;
  text-decoration: none!important;
}

<div
    class="emotion-0"
    id="test-menu"
  >
    <ul
      class="emotion-1"
    >
      <li>
        <a
          class="emotion-2"
          href="#"
        >
          <span
            aria-label="Item 1"
            class="emotion-3"
          >
            Item 1
          </span>
          <span
            class="emotion-4"
          >
            <svg
              fill="none"
              height="6"
              role="img"
              viewBox="0 0 10 6"
              width="10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M.21.22a.69.69 0 0 1 1.01 0L5 4.19 8.78.22a.69.69 0 0 1 1.01 0 .777.777 0 0 1 0 1.06l-4.285 4.5a.69.69 0 0 1-1.01 0L.209 1.28a.777.777 0 0 1 0-1.06Z"
                fill="#81868C"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </a>
        <section
          class="emotion-5"
          style="opacity: 1; height: auto;"
        >
          <ul
            class="emotion-6"
          >
            <li>
              <a
                class="emotion-7"
                href="#"
              >
                <img
                  alt="line"
                  class="emotion-8"
                  src="[object Object]"
                />
                Subitem 1
              </a>
            </li>
          </ul>
        </section>
      </li>
      <li>
        <a
          class="emotion-9"
          href="#"
        >
          <span
            aria-label="Item 2"
            class="emotion-3"
          >
            Item 2
          </span>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
